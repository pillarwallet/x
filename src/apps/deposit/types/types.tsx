export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
