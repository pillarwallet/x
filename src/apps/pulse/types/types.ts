export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
