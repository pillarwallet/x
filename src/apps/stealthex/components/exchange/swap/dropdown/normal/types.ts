
import type React from 'react';
import { StructuredCurrency } from '../../../../../type';
import { CurrencyData } from '../../../../../lib/backend/api';

export type ClickCallback = (currency: CurrencyData) => void;

export type AvailableCurrency =
  | {
    availableCurrency?: Omit<StructuredCurrency, 'fiat'> & {
      all?: CurrencyData[];
    };
    type: 'get';
  }
  | {
    availableCurrency?: StructuredCurrency & {
      all?: CurrencyData[];
    };
    type: 'send';
  };

export type NormalDropdownProps = AvailableCurrency & {
  loading?: boolean;
  selectedCurrency: CurrencyData | null;
  searchContext: string;
  fiat?: boolean;
  onClick?: ClickCallback;
  components?: {
    NotFound?: React.FC<React.PropsWithChildren>;
    Group?: React.FC<React.PropsWithChildren>;
    Item?: React.FC<
      React.PropsWithChildren<
        | { loading: true }
        | {
          loading?: false;
          currency: CurrencyData;
          selected?: boolean;
          onClick?: () => void;
        }
      >
    >;
  };
  groupHeight?: number;
  itemHeight?: number;
  maxHeight?: number;
  embed?: boolean;
};
