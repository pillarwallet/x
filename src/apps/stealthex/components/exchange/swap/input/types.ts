
import { CurrencyData } from '../../../../lib/backend/api';
import type { AvailableCurrency } from '../dropdown/normal/types';

export type SavedAvailableCurrency =
  | {
      floatingDropdown: true;
      availableCurrency: CurrencyData[];
      type: 'get' | 'send';
    }
  | ({
      floatingDropdown?: false;
    } & AvailableCurrency);

export type InputProps = {
  currency: CurrencyData;
  value?: string;
  loading?: boolean;
  loadingCurrency?: boolean;
  error?: string | JSX.Element;
  fiat?: boolean;
  disabled?: boolean;
  withLock?: boolean | null;
  widget?: boolean;
  fixed?: boolean;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onFixedChange?: (fixed: boolean) => void;
  onCurrencySelect?: (currency: CurrencyData) => void;
  pressable?: boolean;
} & SavedAvailableCurrency;
