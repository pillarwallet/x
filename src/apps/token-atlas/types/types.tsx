export type ChainType = {
  chainId: number;
  chainName: string;
};

export enum PeriodFilter {
  HOUR = 'HOUR',
  DAY = 'DAY',
  WEEK = 'WEEK',
  MONTH = 'MONTH',
  YEAR = 'YEAR',
}

export type SelectedTokenType = {
  id: number;
  symbol: string;
  address: string;
  decimals?: number;
  chainId?: number;
  name: string;
  icon?: string;
  price?: number;
};
