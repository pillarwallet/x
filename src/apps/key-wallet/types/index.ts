export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

