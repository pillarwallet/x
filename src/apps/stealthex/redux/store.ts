import { configureStore, TypedStartListening } from '@reduxjs/toolkit';
import exchangeReducer from './reducers/exchange';
import { exchange } from '../lib/backend/api';

export const store = configureStore({
    reducer: {
        exchange: exchangeReducer,
        [exchange.reducerPath]: exchange.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware({ serializableCheck: false }).concat(exchange.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStartListening = TypedStartListening<RootState, AppDispatch>; 