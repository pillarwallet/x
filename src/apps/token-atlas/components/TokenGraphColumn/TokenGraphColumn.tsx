import { sub } from 'date-fns';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';

// hooks
import useAllowedApps from '../../../../hooks/useAllowedApps';
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// reducer
import {
  setPeriodFilter,
  setPriceGraphPeriod,
} from '../../reducer/tokenAtlasSlice';

// utils
import { convertDateToUnixTimestamp } from '../../../../utils/common';
import { limitDigits } from '../../utils/converters';

// types
import {
  MarketHistoryPairData,
  TokenAtlasInfoData,
} from '../../../../types/api';
import { PeriodFilter, SelectedTokenType } from '../../types/types';

// images
import ArrowGreenSmall from '../../images/arrow-circle-green-small.svg';
import ArrowGreen from '../../images/arrow-circle-green.svg';
import ArrowRedSmall from '../../images/arrow-circle-red-small.svg';
import ArrowRed from '../../images/arrow-circle-red.svg';

// components
import SkeletonLoader from '../../../../components/SkeletonLoader';
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import TokenGraph from '../TokenGraph/TokenGraph';
import Body from '../Typography/Body';

type TokenGraphColumnProps = {
  className?: string;
  isLoadingTokenDataInfo: boolean;
};

const TokenGraphColumn = ({
  className,
  isLoadingTokenDataInfo,
}: TokenGraphColumnProps) => {
  const navigate = useNavigate();
  const { setIsAnimated } = useAllowedApps();
  const dispatch = useAppDispatch();
  const tokenDataInfo = useAppSelector(
    (state) => state.tokenAtlas.tokenDataInfo as TokenAtlasInfoData | undefined
  );
  const tokenDataGraph = useAppSelector(
    (state) =>
      state.tokenAtlas.tokenDataGraph as MarketHistoryPairData | undefined
  );
  const periodFilter = useAppSelector(
    (state) => state.tokenAtlas.periodFilter as PeriodFilter
  );
  const selectedToken = useAppSelector(
    (state) => state.tokenAtlas.selectedToken as SelectedTokenType | undefined
  );

  const [viewportWidth, setViewportWidth] = useState<number>(window.innerWidth);
  const [isBrokenImage, setIsBrokenImage] = useState<boolean>(false);

  // The resize handle and listener are to check the viewport size, and change the arrows SVG accordingly
  const handleResize = () => {
    setViewportWidth(window.innerWidth);
  };

  useEffect(() => {
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  const timeFilter = [
    PeriodFilter.HOUR,
    PeriodFilter.DAY,
    PeriodFilter.WEEK,
    PeriodFilter.MONTH,
    PeriodFilter.YEAR,
  ];

  // The percentage showing next to the token price is the price change in the last 24H
  const getArrow = () => {
    if (tokenDataInfo?.price_change_24h) {
      if (tokenDataInfo.price_change_24h > 0) {
        return viewportWidth > 768 ? ArrowGreen : ArrowGreenSmall;
      }
      if (tokenDataInfo.price_change_24h < 0) {
        return viewportWidth > 768 ? ArrowRed : ArrowRedSmall;
      }
    }
    return '';
  };

  // The handleClickTimePeriod makes sure we select the right "from" Unix timestamp to today's Unix timestamp for the price history graph
  const handleClickTimePeriod = (filter: PeriodFilter) => {
    dispatch(setPeriodFilter(filter));
    const now = new Date();
    let from;
    switch (filter) {
      case PeriodFilter.HOUR:
        from = sub(now, { hours: 2 });
        break;
      case PeriodFilter.DAY:
        from = sub(now, { hours: 24 });
        break;
      case PeriodFilter.WEEK:
        from = sub(now, { weeks: 1 });
        break;
      case PeriodFilter.MONTH:
        from = sub(now, { months: 1 });
        break;
      case PeriodFilter.YEAR:
        from = sub(now, { years: 1 });
        break;
      default:
        from = sub(now, { days: 1 });
        break;
    }
    dispatch(
      setPriceGraphPeriod({
        from: convertDateToUnixTimestamp(from),
        to: undefined,
      })
    );
  };

  return (
    <div
      id="token-atlas-token-graph-column"
      className={`flex flex-col ${className} mr-4`}
    >
      <div className="flex flex-col w-full max-w-[460px]">
        <div
          id="token-atlas-graph-column-name-logo-symbol"
          className="flex items-center gap-2 mb-2"
        >
          {isLoadingTokenDataInfo ? (
            <SkeletonLoader $height="29px" $radius="6px" $marginBottom="10px" />
          ) : (
            <>
              <div className="relative w-[30px] h-[30px] rounded-full">
                {tokenDataInfo?.logo && !isBrokenImage ? (
                  <img
                    src={tokenDataInfo.logo}
                    alt="token-logo"
                    className="w-full h-full object-fill rounded-full"
                    data-testid="token-logo-graph-column"
                    onError={() => setIsBrokenImage(true)}
                  />
                ) : (
                  <div className="w-full h-full overflow-hidden rounded-full">
                    <RandomAvatar name={tokenDataInfo?.name || ''} />
                  </div>
                )}

                {/* Overlay text when no token logo available */}
                {(!tokenDataInfo?.logo || isBrokenImage) && (
                  <span className="absolute inset-0 flex items-center justify-center text-lg text-xs font-bold">
                    {tokenDataInfo?.name?.slice(0, 2)}
                  </span>
                )}
              </div>

              <Body className="font-medium text-[27px] mobile:text-[25px]">
                {tokenDataInfo ? tokenDataInfo.name : 'Token not found'}
              </Body>
              <Body className="text-[15px] mobile:text-[13px] text-white_light_grey pt-2">
                {tokenDataInfo?.symbol}
              </Body>
              {tokenDataInfo && (
                <button
                  type="button"
                  className="flex w-fit ml-2 py-3 px-6 text-sm font-semibold uppercase truncate rounded bg-green hover:bg-[#5DE000] text-dark_grey"
                  onClick={() => {
                    setIsAnimated(false);
                    navigate(
                      `/the-exchange?asset=${selectedToken?.name}&blockchain=${selectedToken?.chainId}&address=${selectedToken?.address}`
                    );
                  }}
                >
                  Buy {tokenDataInfo?.symbol}
                </button>
              )}
            </>
          )}
        </div>
        <div
          id="token-atlas-graph-column-price-change"
          className="flex justify-between items-center desktop:items-end"
        >
          {isLoadingTokenDataInfo ? (
            <SkeletonLoader $height="50px" $radius="6px" $marginBottom="10px" />
          ) : (
            <>
              <h1
                id="token-atlas-graph-column-price-today"
                className="text-[60px] mobile:text-[40px] mr-4"
              >
                <span className="text-white_light_grey">$</span>
                {tokenDataInfo?.price && limitDigits(tokenDataInfo.price)}
              </h1>
              <div
                id="token-atlas-graph-column-price-change-percentage"
                className="flex mobile:flex-col tablet:flex-col items-end desktop:mb-5 mb-0"
              >
                {tokenDataInfo?.price_change_24h && (
                  <>
                    <img
                      src={getArrow()}
                      alt="arrow"
                      className={`w-[30px] mr-1 mobile:w-3.5 mobile:mb-2 ${
                        tokenDataInfo.price_change_24h < 0 && 'rotate-180'
                      }`}
                    />
                    <div className="flex">
                      <Body className="text-[15px] mobile:text-[13px]">
                        {tokenDataInfo.price_change_24h.toFixed(3)}
                      </Body>
                      <Body className="text-[11px] font-black mobile:text-[9px] self-start">
                        %
                      </Body>
                    </div>
                  </>
                )}
              </div>
            </>
          )}
        </div>
        <div
          id="token-atlas-graph-column-time-filter"
          className="flex rounded bg-medium_grey max-w-[460px] p-1 gap-1"
        >
          {timeFilter.map((filter, index) => (
            <button
              type="button"
              key={index}
              className={`flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded ${
                tokenDataGraph?.result.data.length
                  ? 'hover:bg-green hover:text-dark_grey'
                  : ''
              } ${periodFilter === filter && tokenDataGraph?.result.data.length ? 'bg-green text-dark_grey' : 'text-white_grey bg-medium_grey'}`}
              onClick={() => handleClickTimePeriod(filter)}
            >
              {filter}
            </button>
          ))}
        </div>
      </div>
      <TokenGraph />
    </div>
  );
};

export default TokenGraphColumn;
