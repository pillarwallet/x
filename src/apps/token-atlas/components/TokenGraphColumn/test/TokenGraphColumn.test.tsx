import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

// redux store
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// provider
import AllowedAppsProvider from '../../../../../providers/AllowedAppsProvider';

// components
import TokenGraphColumn from '../TokenGraphColumn';

// reducer
import {
  setPeriodFilter,
  setTokenDataGraph,
  setTokenDataInfo,
} from '../../../reducer/tokenAtlasSlice';

// types
import {
  MarketHistoryPairData,
  TokenAtlasInfoData,
} from '../../../../../types/api';
import { PeriodFilter } from '../../../types/types';

const mockTokenDataGraph: MarketHistoryPairData = {
  result: {
    data: [
      {
        volume: 1050.25,
        open: 45000.5,
        high: 45500.0,
        low: 44800.25,
        close: 45250.75,
        time: 1712457600, // Unix timestamp
      },
      {
        volume: 980.1,
        open: 45250.75,
        high: 46000.0,
        low: 45000.0,
        close: 45900.0,
        time: 1712544000,
      },
      {
        volume: 1125.8,
        open: 45900.0,
        high: 46250.5,
        low: 45700.0,
        close: 46050.5,
        time: 1712630400,
      },
    ],
  },
};

const mockTokenDataInfo: TokenAtlasInfoData = {
  id: 1,
  market_cap: 100,
  market_cap_diluted: 100,
  liquidity: 150,
  price: 105,
  off_chain_volume: 150,
  volume: 150,
  volume_change_24h: 1.54,
  volume_7d: 2.3,
  is_listed: true,
  price_change_24h: 0.44,
  price_change_1h: 0.1,
  price_change_7d: 1.8,
  price_change_1m: 3.4,
  price_change_1y: 6.7,
  ath: 146,
  atl: 96,
  name: 'TOKEN',
  symbol: 'TKN',
  logo: 'tokenLogo.png',
  rank: 1047,
  contracts: [],
  total_supply: '300',
  circulating_supply: '170',
};

describe('<TokenGraphColumn />', () => {
  beforeEach(() => {
    store.dispatch(setTokenDataInfo(mockTokenDataInfo));
    store.dispatch(setTokenDataGraph(mockTokenDataGraph));
    store.dispatch(setPeriodFilter(PeriodFilter.DAY));
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <AllowedAppsProvider>
              <TokenGraphColumn isLoadingTokenDataInfo={false} />
            </AllowedAppsProvider>
          </MemoryRouter>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('displays token information correctly', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <AllowedAppsProvider>
            <TokenGraphColumn isLoadingTokenDataInfo={false} />
          </AllowedAppsProvider>
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByTestId('token-logo-graph-column')).toBeInTheDocument();
    expect(screen.getByText(mockTokenDataInfo.name)).toBeInTheDocument();
    expect(screen.getByText(mockTokenDataInfo.symbol)).toBeInTheDocument();
    expect(screen.getByText(`${mockTokenDataInfo.price}`)).toBeInTheDocument();
    expect(
      screen.getByText(`${mockTokenDataInfo.price_change_24h.toFixed(3)}`)
    ).toBeInTheDocument();
  });
});
