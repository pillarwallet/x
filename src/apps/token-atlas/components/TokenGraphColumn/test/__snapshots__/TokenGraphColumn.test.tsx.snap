// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenGraphColumn /> renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col undefined mr-4"
  id="token-atlas-token-graph-column"
>
  <div
    className="flex flex-col w-full max-w-[460px]"
  >
    <div
      className="flex items-center gap-2 mb-2"
      id="token-atlas-graph-column-name-logo-symbol"
    >
      <div
        className="relative w-[30px] h-[30px] rounded-full"
      >
        <img
          alt="token-logo"
          className="w-full h-full object-fill rounded-full"
          data-testid="token-logo-graph-column"
          onError={[Function]}
          src="tokenLogo.png"
        />
      </div>
      <p
        className="text-[13px] font-semibold font-medium text-[27px] mobile:text-[25px]"
      >
        TOKEN
      </p>
      <p
        className="text-[13px] font-semibold text-[15px] mobile:text-[13px] text-white_light_grey pt-2"
      >
        TKN
      </p>
      <button
        className="flex w-fit ml-2 py-3 px-6 text-sm font-semibold uppercase truncate rounded bg-green hover:bg-[#5DE000] text-dark_grey"
        onClick={[Function]}
        type="button"
      >
        Buy 
        TKN
      </button>
    </div>
    <div
      className="flex justify-between items-center desktop:items-end"
      id="token-atlas-graph-column-price-change"
    >
      <h1
        className="text-[60px] mobile:text-[40px] mr-4"
        id="token-atlas-graph-column-price-today"
      >
        <span
          className="text-white_light_grey"
        >
          $
        </span>
        105
      </h1>
      <div
        className="flex mobile:flex-col tablet:flex-col items-end desktop:mb-5 mb-0"
        id="token-atlas-graph-column-price-change-percentage"
      >
        <img
          alt="arrow"
          className="w-[30px] mr-1 mobile:w-3.5 mobile:mb-2 false"
          src={
            {
              "ReactComponent": {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "default": "arrow-circle-green.svg",
            }
          }
        />
        <div
          className="flex"
        >
          <p
            className="text-[13px] font-semibold text-[15px] mobile:text-[13px]"
          >
            0.440
          </p>
          <p
            className="text-[13px] font-semibold text-[11px] font-black mobile:text-[9px] self-start"
          >
            %
          </p>
        </div>
      </div>
    </div>
    <div
      className="flex rounded bg-medium_grey max-w-[460px] p-1 gap-1"
      id="token-atlas-graph-column-time-filter"
    >
      <button
        className="flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded hover:bg-green hover:text-dark_grey text-white_grey bg-medium_grey"
        onClick={[Function]}
        type="button"
      >
        HOUR
      </button>
      <button
        className="flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded hover:bg-green hover:text-dark_grey bg-green text-dark_grey"
        onClick={[Function]}
        type="button"
      >
        DAY
      </button>
      <button
        className="flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded hover:bg-green hover:text-dark_grey text-white_grey bg-medium_grey"
        onClick={[Function]}
        type="button"
      >
        WEEK
      </button>
      <button
        className="flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded hover:bg-green hover:text-dark_grey text-white_grey bg-medium_grey"
        onClick={[Function]}
        type="button"
      >
        MONTH
      </button>
      <button
        className="flex-1 text-[11px] font-semibold capitalize truncate py-3 rounded hover:bg-green hover:text-dark_grey text-white_grey bg-medium_grey"
        onClick={[Function]}
        type="button"
      >
        YEAR
      </button>
    </div>
  </div>
  <div
    className="flex w-[99%] mb-20 h-full max-h-[400px] mobile:mb-0"
    id="token-atlas-token-graph"
  >
    <canvas
      data-testid="price-graph"
      height={150}
      role="img"
      width={300}
    />
  </div>
</div>
`;
